sklearn/utils/_array_api.py

```diff
@@ -74,22 +75,35 @@ def _check_array_api_dispatch(array_api_dispatch):
             )
 
 
-def device(x):
+def device(*array_list):
     """Hardware device the array data resides on.
 
     Parameters
     ----------
-    x : array
-        Array instance from NumPy or an array API compatible library.
+    *array_list : arrays
+        List of array instances from NumPy or an array API compatible library.
 
     Returns
     -------
     out : device
         `device` object (see the "Device Support" section of the array API spec).
     """
-    if isinstance(x, (numpy.ndarray, numpy.generic)):
-        return "cpu"
-    return x.device
+    if not array_list:
+        raise ValueError("At least one input array expected, got none.")
+
+    devices = set()
+    for array in array_list:
+        if isinstance(array, (numpy.ndarray, numpy.generic)) or not hasattr(
+            array, "device"
+        ):
+            devices.add("cpu")
+        else:
+            devices.add(array.device)
+
+    if len(devices) > 1:
+        raise ValueError("Input arrays use different devices.")
```

Comment:
Let's make this message more informative.

```suggestion
        raise ValueError(f"Input arrays use different devices: {', '.join(sorted(devices))}")
```

It might be challenging to test for this case with our current CI without a GPU device so I am fine with leaving this branch uncovered by tests.